wonnx/src/gpu.rs

```diff
@@ -657,46 +661,64 @@ impl GpuTensor {
         queue: &wgpu::Queue,
     ) -> Result<OutputTensor, GpuError> {
         let buffer_slice = self.buffer.slice(..);
+        let shape = self.shape.clone();
 
         // On wgpu we can MAP_READ a buffer that is also used as STORAGE, but WebGPU (on at least Chrome)
         // disallows this. Therefore we need to do an additional copy into a MAP_READ buffer when reading back a
         // STORAGE buffer when on WebGPU.
         #[cfg(target_arch = "wasm32")]
-        let output_data = wgpu::util::DownloadBuffer::read_buffer(device, queue, &buffer_slice)
-            .await
-            .unwrap();
+        {
+            let (tx, rx) = futures_intrusive::channel::shared::oneshot_channel();
+
+            wgpu::util::DownloadBuffer::read_buffer(device, queue, &buffer_slice, move |b| {
+                // Called on download completed
+                tx.send(match b {
+                    Ok(b) => Ok(Self::read_bytes_to_vec(&b, shape)),
+                    Err(abe) => Err(GpuError::BufferAsyncError(abe)),
+                })
+                .unwrap();
+            });
+            device.poll(wgpu::Maintain::Wait);
+            // The callback will have been called by now due to poll(Wait)
+            rx.receive().await.unwrap()
+        }
 
         #[cfg(not(target_arch = "wasm32"))]
-        let output_data = {
-            let _ = queue; // Need this because otherwise compiler complains we are not using the queue parameter
-            let buffer_future = buffer_slice.map_async(wgpu::MapMode::Read);
-            device.poll(wgpu::Maintain::Wait);
-            buffer_future.await.expect("failed to run compute on gpu!");
-            buffer_slice.get_mapped_range()
-        };
+        {
+            let output_data = {
+                let _ = queue; // Need this because otherwise compiler complains we are not using the queue parameter
+                buffer_slice.map_async(wgpu::MapMode::Read, |_| {});
+                device.poll(wgpu::Maintain::Wait);
+                buffer_slice.get_mapped_range()
+            };
 
+            let result = Self::read_bytes_to_vec(&output_data[..], shape);
+            drop(output_data);
+            self.buffer.unmap();
+            Ok(result)
+        }
```

Comment:
```suggestion
            let (tx, rx) = futures_intrusive::channel::shared::oneshot_channel();
            buffer_slice.map_async(wgpu::MapMode::Read, move |v| {
                tx.send(v).unwrap();
            });
            device.poll(wgpu::Maintain::Wait);

            // Awaits until `buffer_future` can be read from
            if let Some(Ok(())) = rx.receive().await {
                let output_data = { buffer_slice.get_mapped_range() };

                let result = Self::read_bytes_to_vec(&output_data[..], shape);
                drop(output_data);
                self.buffer.unmap();
                Ok(result)
            } else {
                Err(GpuError::OutputReadError())
            }
            buffer_slice.map_async(wgpu::MapMode::Read, move |v| {
                tx.send(v).unwrap();
            });
            device.poll(wgpu::Maintain::Wait);

            // Awaits until `MapRead` is effective
            if let Some(Ok(())) = rx.receive().await {
                let output_data = { buffer_slice.get_mapped_range() };

                let result = Self::read_bytes_to_vec(&output_data[..], shape);
                drop(output_data);
                self.buffer.unmap();
                Ok(result)
            } else {
                Err(GpuError::OutputReadError())
            }
```
The `map_async` does not give you any guarantee that its effective without the callback function.

Reply:
Good catch, although not an issue right now (due to the current implementation) you are technically right and this behaviour could change :+1: